scanpy/get.py

```diff
@@ -227,10 +247,26 @@ def var_df(
             " in `adata.obs_names`."
         )
 
-    # Make df
-    df = pd.DataFrame(index=adata.var_names)
-    for k, l in zip(keys, lookup_keys):
-        df[k] = adata.var_vector(l, layer=layer)
+    # initialize df
+    df = pd.DataFrame(index=adata.var.index)
+
+    # add obs values
+    if len(obs_names) > 0:
+        X = _get_obs_rep(adata, layer=layer)
+        matrix = X[adata.obs_names.get_indexer(obs_names), :]
```

Comment:
Just had a thought, I don't think this will work if X is backed (as a dense array) and `obs_names` isn't sorted. `h5py.Dataset`s require that the indices be in order. This should probably get a test case.

An alternative is `_get_obs_rep(adata[obs_names], ...).copy()`, but this will have performance issues with `raw`.

Reply:
Internally in `anndata`, I index in-order then reorder the array for `h5py.Dataset` objects.

Reply:
How can we test for this? I added a test to read a backed dataset from the included datasets. Why this works? or by chance are the indices ordered?

Reply:
The indices look ordered, since you're getting them like this: `list(adata.var_names[:10])`.

I think `list(adata.var_names[10::-1]) would be enough to make it fail. In `AnnData` valid indices in random order are generated with functions in `anndata/tests/helpers.py`.